api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,131 @@
+                                                                      uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
+ * @param root_context_id is an identifier for one or more related plugins.
+ * @param plugin_configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" OnConfigureResult proxy_on_configure(uint32_t root_context_id,
+                                                uint32_t plugin_configuration_size);
+
+// Stream calls.
+
+/**
+ * Called when a request, stream or other ephemeral context is created.
+ * @param context_id is an identifier of the ephemeral context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ */
+extern "C" void proxy_on_context_create(uint32_t context_id, uint32_t root_context_id);
+
+// Stream and Non-stream calls.
+
+/**
+ * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
+ * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
+ * called when the VM is going to shutdown.
+ * @param context_id is an identifier the context.
+ * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
+ * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
+ * future proxy_done() call by the root context.
+ */
+extern "C" OnDoneResult proxy_on_done(uint32_t context_id);
+
+/**
+ * Called when the context is being deleted and will no longer receive any more calls.
+ * @param context_id is an identifier the context.
+ */
+extern "C" void proxy_on_delete(uint32_t context_id);
+
+//
+// ABI calls from the VM to the host.
+//
+
+// Configuration and Status
+
+/**
+ * Called from the VM to get any configuration. Valid only when in proxy_on_start() (where it will
+ * return a VM configuration), proxy_on_configure() (where it will return a plugin configuration) or
+ * in proxy_validate_configuration() (where it will return a VM configuration before
+ * proxy_on_start() has been called and a plugin configuration after).
+ * @param configuration_ptr a pointer to a location which will be filled with either nullptr (if no
+ * configuration is available) or a pointer to a malloc()ed block containing the configuration
+ * bytes.
+ * @param configuration_size a pointer to a location containing the size (or zero) of any returned
+ * configuration byte block.
+ * @return a WasmResult: OK, InvalidMemoryAccess. Note: if OK is returned  *configuration_ptr may
+ * be nullptr.
+ */
+extern "C" WasmResult proxy_get_configuration(const char** configuration_ptr,
+                                              size_t* configuration_size);
```

Comment:
Sorry I still don't understand this API. In the other APIs, you pass the size of the configuration. If this API gets filled with memory and the size, why is the size passed in the other API? Who is actually doing the malloc here? The host?

Reply:
The host is doing the malloc.  We pass the size because if the size is zero you could just not get the configuration, and for consistency since we want to pass the size for buffers since the ABI is to request a range and the code may want to do something different (i.e. request blocks) if the size is very large.

Reply:
How is it possible with this API to request a range? Shouldn't `configuration_size` be in in/out param then, and shouldn't there by an offset param? Can we go ahead and add that if that is the case?

Reply:
This particular ABI doesn't request a range.  As I said, we do this for consistency in this ABI since we want to pass the size for buffers since the buffer ABI is to request a range.

Matching this to the existing buffer ABI in envoyproxy/proxy-wasm.  PTAL